pkg/tasks/tasks.go

```diff
@@ -39,23 +41,23 @@ func NewTaskRunner(client *client.Client, taskGroups ...*TaskGroup) *TaskRunner
 
 // RunAll executes all registered task groups sequentially. For each group the
 // taskGroup.RunConcurrently function is called.
-func (tl *TaskRunner) RunAll() (string, error) {
+func (tl *TaskRunner) RunAll() TaskGroupErrors {
 	for i, tGroup := range tl.taskGroups {
 		klog.V(2).Infof("processing task group %d of %d", i+1, len(tl.taskGroups))
-
-		if name, err := tGroup.RunConcurrently(); err != nil {
-			return name, err
+		tErrors := tGroup.RunConcurrently()
+		if len(tErrors) > 0 {
+			return tErrors
 		}
-
 	}
-	return "", nil
+	return nil
 }
 
 // RunConcurrently dispatches all tasks in a task group. The tasks are scheduled
-// concurrently. Returns the first error if any are encountered.
-func (tg *TaskGroup) RunConcurrently() (string, error) {
+// concurrently. Returns all the errors that are encountered.
+func (tg *TaskGroup) RunConcurrently() TaskGroupErrors {
 	var g errgroup.Group
 	tgLength := len(tg.tasks)
+	errChan := make(chan TaskErr, len(tg.tasks))
```

Comment:
```suggestion
	errChan := make(chan TaskErr, len(tg.tasks))
	defer close(errChan)
```
Maybe this makes sense here instead of the explicit close further down?

Reply:
This does not work as the range keeps receiving values until the channel is closed. On doing this, the for loop never ends.

https://tour.golang.org/concurrency/4

Reply:
Ahh I see, thanks.

Should we then have a comment further down, that `g.Wait` and `close(errChan)` must be called before we receive from `errChan` with `range`?

Reply:
Done, add comment

Reply:
Same line, different issue. Why not 
```suggestion
	errChan := make(chan TaskErr, tgLength)
```

Reply:
done